Mathlib/Algebra/Group/ForwardDiff.lean

```diff
@@ -193,3 +193,197 @@ lemma fwdDiff_addChar_eq {M R : Type*} [AddCommMonoid M] [Ring R]
+A polynomial `P(x) = ∑_{k=0..n} aₖ xᵏ` has `Δ^[n+1] P = 0`.
+-/
+theorem fwdDiff_iter_succ_sum_eq_zero {n : ℕ} (a : ℕ → R):
+    ((fwdDiffₗ R R 1 ^ (n + 1)) fun x ↦ ∑ k ∈ Finset.range (n + 1), a k * (x ^ k)) = 0 := by
+  induction' n with n ih
+  · unfold fwdDiffₗ
+    simp; ext x
+    simp only [Pi.zero_apply]
+  · rw [pow_succ, Module.End.mul_apply]
+    have :((fwdDiffₗ R R 1 ^ (n + 1)) (fun x => ∑ k ∈ range (n + 1), a k * x ^ k)) =
+      ∑ k ∈ range (n + 1), (fwdDiffₗ R R 1 ^ (n + 1))
+      ((fun x : R ↦ a k) * (fun x : R ↦  x ^ k)) := by
+      ext x; simp
+      have : (fun x => ∑ k ∈ range (n + 1), a k * x ^ k) =
+        (∑ k ∈ range (n + 1), fun x =>  a k * x ^ k) := by
+        ext x; simp only [Finset.sum_apply]
+      simp only [this, map_sum, sum_apply]
+      rfl
+    rw [this] at ih
+    have : ((fwdDiffₗ R R 1) fun x ↦ ∑ k ∈ range (n + 1 + 1), a k * x ^ k) =
+      ∑ k ∈ range (n + 1 + 1), a k • (fun x : R ↦  (x + 1) ^ k) -
+      ∑ k ∈ range (n + 1 + 1),a k • fun x : R ↦ x ^ k := by
+      ext x; simp
+      unfold fwdDiff
+      rfl
+    rw [this]
+    simp only [map_sub, map_sum, coe_fwdDiffₗ_pow]
+    ext x
+    rw [Finset.sum_range_succ]
+    nth_rw 2 [Finset.sum_range_succ]
+    simp only [fwdDiff_iter_const_smul, Pi.sub_apply, Pi.add_apply, sum_apply, Pi.smul_apply,
+      smul_eq_mul]
+    rw [← add_sub, ← coe_fwdDiffₗ_pow]
+    have :  ((fwdDiffₗ R R 1 ^ (n + 1)) fun x => x ^ (n + 1)) (x + 1) =
+        ((fwdDiffₗ R R 1 ^ (n + 1)) fun x => (x + 1) ^ (n + 1)) x := by
+      simp only [coe_fwdDiffₗ_pow, fwdDiff_iter_eq_sum_shift, fwdDiff_iter_eq_sum_shift,
+        Int.reduceNeg, nsmul_eq_mul, mul_one, zsmul_eq_mul, Int.cast_mul, Int.cast_pow,
+        Int.cast_neg, Int.cast_one, Int.cast_natCast,
+        add_assoc (b := (1 : R)), add_comm (a := (1 : R)), ← add_assoc]
+    simp only [← this, fwdDiff_iter_eq_factorial (n := n + 1) (R := R), sub_add_cancel_right, ←
+      sub_eq_add_neg, ← Finset.sum_sub_distrib]
+    exact Finset.sum_eq_zero fun k hk ↦ have _ := Finset.mem_range.1 hk; by
+      rw [← mul_sub, fwdDiff_iter_pow_eq_zero_of_lt (by linarith) ]
+      simp only [Pi.zero_apply, sub_zero]
+      have : (fwdDiffₗ R R 1 ^ (n + 1)) (fun x => (x + 1) ^ k) x =
+      (fwdDiffₗ R R 1 ^ (n + 1)) (fun x => x ^ k) (x + 1) := by
+        rw [coe_fwdDiffₗ_pow, fwdDiff_iter_eq_sum_shift, fwdDiff_iter_eq_sum_shift]
+        congr 1; ext k
+        simp only [Int.reduceNeg, nsmul_eq_mul, mul_one,
+          zsmul_eq_mul, Int.cast_mul, Int.cast_pow, Int.cast_neg, Int.cast_one, Int.cast_natCast]
+        ring
+      rw [this, fwdDiff_iter_pow_eq_zero_of_lt (by linarith)]
+      simp only [Pi.zero_apply, mul_zero]
+
+/--
+**Newton's series** for a polynomial function.
+Any function `f` defined by a polynomial can be expressed as a sum of its forward
+differences at `0`, weighted by binomial coefficients.
+`f(x) = ∑_{k=0..p} (p choose k) * (Δ^k f)(0)`.
+-/
+theorem fwdDiffTab_0th_diag_poly' {n p : ℕ} (a : ℕ → R):
```

Comment:
I'm not seeing anything called `fwdDiffTab` in the signature, so this probably needs another name. Also add a space before the final `:`